File: src/insertion_sort.py
import random
import copy


def insertion_sort(array):
    """
    Complexity is O(n*n)
    """
    for j in range(1, len(array)):
        key = array[j]  # the key is import
        i = j - 1
        while i >= 0 and array[i] > key:
            array[i + 1] = array[i]
            i -= 1
        array[i + 1] = key
    return array


if __name__ == "__main__":
    array_1 = range(10)
    array_2 = range(-10, 10)
    array_3 = range(-10, 0)
    for array in [array_1, array_2, array_3]:
        original = copy.copy(array)
        random.shuffle(array)
        assert insertion_sort(array) == original

File: src/isort/insertion_sort.go
package main

import (
    "utils"
    "fmt"
)

func insertion_sort(array []int) {
    for j, _ := range array {
        key := array[j]
        i := j-1
        for ; i>=0 && array[i] > key; i-=1 {
            array[i+1] = array[i]
        }
        array[i+1] = key
    }
}

func main(){
    array_1 := utils.GetIntSlice(10)
    array_2 := utils.GetIntSlice(10000)
    utils.PrintRunningTime(insertion_sort, array_1)
    utils.PrintRunningTime(insertion_sort, array_2)
    fmt.Println(array_1)
    fmt.Println(array_2)
}
